lib/galaxy/tools/execute.py

```diff
@@ -110,6 +116,12 @@ def execute_single_job(execution_slice, completed_job):
     else:
         # Make sure collections, implicit jobs etc are flushed even if there are no precreated output datasets
         trans.sa_session.flush()
+
+    if job_datasets:
+        for job, datasets in job_datasets.items():
+            for dataset_instance in datasets:
+                dataset_instance.dataset.job_id = job.id
```

Comment:
```suggestion
    if job_datasets:
        for job, datasets in job_datasets.items():
            for dataset_instance in datasets:
                dataset_instance.dataset.job = job
```
ideally we haven't flushed the job yet at this point, so it's better to just add the job object directly and let the session resolve this during the flush